File: leetcode/python/0207. Course Schedule.py
# Time complexity: O(V + E), where V is the number of courses, E is the number of edges
# Space complexity: O(V + E)

class Solution(object):
    def canFinish(self, numCourses, prerequisites):
        visited = [0 for _ in range(numCourses)]
        recstack = [0 for _ in range(numCourses)]
        adj_list = {x:[] for x in range(numCourses)}
        
        for pre,post in prerequisites:
            if pre in adj_list:
                adj_list[pre].append(post)
        
        for node in adj_list.keys():
            if visited[node] == False:
                if self.findCycle(node, visited, recstack, adj_list):
                    return False
        return True
        

    def findCycle(self, node, visited, recstack, adjList):
        visited[node] = 1   
        recstack[node] = 1

        for n in adjList[node]:
            if visited[n] == 0:
                if self.findCycle(n, visited, recstack, adjList) == True:
                    return True

            elif recstack[n] == 1:
                return True
        
        recstack[node] = 0
        return False
        

File: leetcode/java/0207. Course Schedule.java
package leetcode.java;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

// Time complexity: O(V + E), where V is the number of courses, E is the number of edges
// Space complexity: O(V + E)

class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        if (prerequisites.length == 0) {
            return true;
        }
        
        HashMap<Integer, List<Integer>> adjList = this.makeAdjList(numCourses, prerequisites);
        int[] visited = new int[numCourses];

        for (int i=0; i<numCourses; i++) {
            if (!this.dfs(adjList, visited, i)) {
                return false;
            }
        }

        return true;

    }

    public HashMap<Integer, List<Integer>> makeAdjList(int numCourses, int[][] prerequisites) {
        HashMap<Integer, List<Integer>> res = new HashMap<>();
        
        for (int i=0; i< numCourses; i++) {
            res.put(i, new ArrayList<>());
        }

        for (int[] prereq : prerequisites) {
            int to = prereq[1];
            int from = prereq[0];

            List<Integer> temp = res.get(to);
            temp.add(from);
            res.put(to, temp);
        }
        return res;
    }

    public boolean dfs(HashMap<Integer, List<Integer>> adjList, int[] visited, int source) {
        
        if (visited[source] == 1) {
            return false;
        }

        visited[source] = 1;

        for (int n : adjList.get(source)) {
            if (!this.dfs(adjList, visited, n)) {
                return false;
            }
        }

        adjList.put(source,new ArrayList<>());
        visited[source] = 0;
        return true;
    }
}